osquery/tables/networking/windows/routes.cpp

```diff
@@ -0,0 +1,168 @@
+      returnMapping.insert(
+          std::make_pair(currentRow.InterfaceIndex, currentRow));
+    }
+  }
+
+  return returnMapping;
+}
+
+QueryData genRoutes(QueryContext& context) {
+  QueryData results;
+  PMIB_IPFORWARD_TABLE2* ipTable = nullptr;
+
+  ipTable = static_cast<PMIB_IPFORWARD_TABLE2*>(
+      malloc(sizeof(PMIB_IPFORWARD_TABLE2)));
+  auto result = GetIpForwardTable2(AF_UNSPEC, ipTable);
+
+  if (result != NO_ERROR) {
+    FreeMibTable(ipTable);
+
+    return results;
+  }
+
+  auto numEntries = ipTable[0]->NumEntries;
+  auto interfaces = getInterfaceRowMapping();
+  auto adapters = getAdapterAddressMapping();
+
+  for (unsigned long i = 0; i < numEntries; ++i) {
+    Row r;
+    std::string interfaceIpAddress;
+    PVOID ipAddress = nullptr;
+    PVOID gateway = nullptr;
+    auto currentRow = ipTable[0]->Table[i];
+    auto addrFamily = currentRow.DestinationPrefix.Prefix.si_family;
+    auto actualInterface = interfaces.at(currentRow.InterfaceIndex);
+    if (addrFamily == AF_INET6) {
+      std::vector<char> buffer(INET6_ADDRSTRLEN);
+
+      r["mtu"] = INTEGER(actualInterface.NlMtu);
+      // These are all technically "on-link" addresses according to
+      // `route print -6`.
+      r["type"] = "local";
+      auto ipAddress = std::make_unique<IN6_ADDR>(
+          currentRow.DestinationPrefix.Prefix.Ipv6.sin6_addr);
+      auto gateway =
+          std::make_unique<IN6_ADDR>(currentRow.NextHop.Ipv6.sin6_addr);
+
+      InetNtop(addrFamily, ipAddress.get(), buffer.data(), buffer.size());
+      r["destination"] = SQL_TEXT(buffer.data());
+      InetNtop(addrFamily, gateway.get(), buffer.data(), buffer.size());
+      r["gateway"] = SQL_TEXT(buffer.data());
+    } else if (addrFamily == AF_INET) {
+      std::vector<char> buffer(INET_ADDRSTRLEN);
+      auto ipAddress = std::make_unique<IN_ADDR>(
+          currentRow.DestinationPrefix.Prefix.Ipv4.sin_addr);
+      auto gateway =
+          std::make_unique<IN_ADDR>(currentRow.NextHop.Ipv4.sin_addr);
+
+      InetNtop(addrFamily, ipAddress.get(), buffer.data(), buffer.size());
+      r["destination"] = SQL_TEXT(buffer.data());
+      InetNtop(addrFamily, gateway.get(), buffer.data(), buffer.size());
+      r["gateway"] = SQL_TEXT(buffer.data());
+
+      // The software loopback is not returned by GetAdaptersInfo, so any
+      // lookups into that index must be skipped and default values set.
+      PIP_ADAPTER_INFO actualAdapter = nullptr;
+      if (currentRow.InterfaceIndex != 1) {
+        actualAdapter = adapters.at(currentRow.InterfaceIndex).get();
+        interfaceIpAddress = actualAdapter->IpAddressList.IpAddress.String;
+        r["mtu"] = INTEGER(actualInterface.NlMtu);
+      } else {
+        interfaceIpAddress = "127.0.0.1";
+        r["mtu"] = UNSIGNED_BIGINT(0xFFFFFFFF);
+      }
+      r["type"] = currentRow.Loopback ? "local" : "remote";
+    }
+    r["interface"] = SQL_TEXT(interfaceIpAddress);
```

Comment:
Aye, that is how the `route` command on Windows works. E.G.
```
C:\> route print -4
...
IPv4 Route Table
===========================================================================
Active Routes:
Network Destination        Netmask          Gateway       Interface  Metric
          0.0.0.0          0.0.0.0     172.25.164.1   172.25.165.115     25
```

An interface can also have a human readable name, or a MAC.